apps/labs/posts/scikit-lego-narwhals.md

```diff
@@ -0,0 +1,115 @@
+the above:
+
+4. if it's Polars, then perform Polars dataframe operations
+
+This looks simple, but it actually opens up a rabbit hole of extra steps, such as:
+
+5. if it's Modin, ...
+6. if it's cuDF, ...
+7. if it's fireducks, ...
+
+In order to address all the above, scikit-lego opted for a simpler approach: use Narwhals.
+The steps then become:
+
+1. run [`narwhals.from_native`](https://narwhals-dev.github.io/narwhals/api-reference/narwhals/#narwhals.from_native) on the input
+2. if it's NumPy, then perform array operations
+3. otherwise, express dataframe logic using the subset of the Polars API supported by Narwhals
+
+Like this, Narwhals takes care of translating its (Polars-like) syntax to each
+supported backend. Read the [Narwhals docs](https://narwhals-dev.github.io/narwhals) for more
+info on how to use it, and for a tutorial.
+
+## Why not "just convert to pandas"?
+
+There are different ways of converting non-pandas input to pandas. Rather than using Narwhals,
+couldn't scikit-lego just have done something like:
+
+- if the input is pandas, keep as-is
+- if the input is a non-pandas dataframe, then convert to pandas
+
+However, even in the best-case zero-copy-conversion scenario, this presents several issues:
+
+- once the input has been converted to pandas, there's no standardised way of converting back to
+  the original dataframe library.
+- using pandas as an engine may be significantly slower than doing everything using the original
+  dataframe
+
+Furthermore, converting to pandas may present a cost - for example, if you start with a Polars
+LazyFrame, which can run significantly faster because it delays reading data into memory and performs
+[several optimisations](https://docs.pola.rs/user-guide/lazy/optimizations/),
+then you're required to call [`.collect`](https://docs.pola.rs/py-polars/html/reference/lazyframe/api/polars.LazyFrame.collect.html) on it before converting to pandas.
+
+To drive the second point home, here are some timings comparing:
+
+- running `add_lags` on a 10-million-row Polars dataframe directly
+- converting that same dataframe to pandas and running `add_lags` on that
+
+![with native Polars support `add_lags` takes ~ 11 ms, whereas going via pandas it
+takes ~ 1470 ms. Image produced with `great-tables` package](/posts/scikit-lego-narwhals/comparison.png)
+
+[Code to reproduce](https://gist.github.com/MarcoGorelli/1da1971063caf0b3e5133f5dfba3315b).
+
+Note that we're dealing with purely numeric data here, so the conversion is extremely cheap.
+The reason that the second approach takes over 100x longer is that Polars generally excels at
+feature engineering. As the size of the dataframe increases, the comparison
+scales even better in the "native Polars support" direction.
+
+That's why "just convert to pandas" is a disappointing solution, and why it's worth it to
+invest the time to achieve dataframe-agnosticism.
+
+## Will the future of data science become "BYODF" (bring your own dataframe)?
```

Comment:
```suggestion
## Will the future of data science become BYOD (bring your own dataframe)?
```

Reply:
that might be more technically correct, but the great-tables blog post which introduced (as far as I know) the "byodf" acronym wrote it like that